Initial Classes

mecha opened the pull request.

mecha committed:
Fixed autoload namespace

mecha committed:
Aliased `develop` branch to `0.1.x-dev`

mecha committed:
Added test dirs to dev autoload

mecha committed:
Added test stub dir to autoload-dev

mecha committed:
Updated dependencies and test config

mecha committed:
Added `PdoAwareTrait`

mecha committed:
Added `ExecutePdoQueryCapableTrait`

mecha committed:
Added `AbstractPdoResourceModel`

mecha committed:
Imported traits from old abstract package

mecha committed:
Added database test case stub class

mecha committed:
Added `PdoSelectResourceModel`

mecha committed:
Added `PdoInsertResourceModel`

mecha committed:
Added `PdoUpdateResourceModel`

mecha committed:
Added `PdoDeleteResourceModel`

mecha committed:
Now depending on `^0.1` interface package

XedinUnknown commented on src/DeleteCapablePdoTrait.php:
```diff
@@ -0,0 +1,107 @@
+     */
+    protected function _delete(LogicalExpressionInterface $condition = null)
+    {
+        $fieldNames   = $this->_getSqlDeleteFieldNames();
```
Here, and in other similar cases, could you explain what the return value is, and why it is necessary?

XedinUnknown replied:
Also, why are they ignored when building the hash map?

mecha replied:
The return value is documented in the doc for `_getSqlDeleteFieldNames()` as `string[]|Stringable[] A list of field names`. Is this unclear?

They are ignored because the value-hash map is for **values**. Field names are not values. Consider an expression that would be rendered into `WHERE name = "duck"`. To invoke this query with PDO, only `"duck"` needs to be extracted and replaced with a hash.

XedinUnknown replied:
I understand what the _type_ is. But I don't understand *what* it is, why it is there, what purpose it serves, and how "field names" can be matched with the idea of "deleting" something.

mecha replied:
A query like this:

```
DELETE FROM `t` WHERE `name` = "duck" OR `name` = "duckling";
```

Needs to be transformed into this:

```
DELETE FROM `t` WHERE `name` = :12345 OR `name` = :67890;
```

Since the WHERE clause is expressed as a logical expression, the expression must be traversed, and each value encountered must have a hash generated for it. This is what the value-hash map is. Since expressions are immutable, the expression itself is not changed - but the value-hashmap is later given to the expression renderer.

Generating the value-hashmap is not so straight-forward, because expressions can be literally anything. Therefore, the algorithm accepts a list of strings to ignore. In this case, things like "name" are ignored. Because otherwise, this would be the resulting query, which is incorrect:

```
DELETE FROM `t` WHERE :123abc = :345def OR :567abc = :789def;
```

XedinUnknown commented on src/GetPdoExpressionHashMapCapableTrait.php:
```diff
@@ -0,0 +1,79 @@
+
+        if ($condition instanceof LiteralTermInterface) {
+            $value = $condition->getValue();
+            $value = $this->_normalizeString($value);
```
Undocumented throw?

XedinUnknown replied:
Also, looks like a case for `OutOfRangeException`.

mecha replied:
Yep! Agree.

XedinUnknown commented on src/GetPdoExpressionHashMapCapableTrait.php:
```diff
@@ -0,0 +1,79 @@
+     */
+    protected function _getPdoExpressionHashMap(TermInterface $condition, array $ignore = [], array &$valueHashMap = [])
+    {
+        if ($valueHashMap === null) {
```
How about defaulting this to `null`, and not populating it in that case?

XedinUnknown replied:
More importantly, why both accept it by reference, and return it?

mecha replied:
It's accepted by reference for performance reasons. Since it's recursive, the resulting arrays would have to be merged. By passing the result by reference, no merging is needed.

But the parameter is also optional. The caller need not pass an existing map to get a resulting map. So the result is returned as well.

It might seem redundant since the caller can already retrieve the result from the argument passed by reference, but since this method had no other viable return value, I decided it just makes sense to return the map - because it _is_ the result after all.

Not populating and returning a pdo expression hash map in a method literally called `_getPdoExpressionHashMap` doesn't make sense to me.

XedinUnknown replied:
Re-naming to something like `_populatePdoExpressionHashMap()` could solve the problem.

mecha replied:
And what _is_ the problem?

XedinUnknown commented on src/GetPdoValueHashStringCapableTrait.php:
```diff
@@ -0,0 +1,39 @@
+     */
+    protected function _getPdoValueHashString($value)
+    {
+        return ':' . hash('crc32b', $this->_normalizeString($value));
```
Info: support for the `crc32b` hashing algorithm was added in PHP 5.1.2, and is continued through to 7.2.2, which is the current latest stable.

XedinUnknown commented on src/InsertCapablePdoTrait.php:
```diff
@@ -0,0 +1,236 @@
+
+        $result = [];
+
+        foreach ($this->_getSqlInsertFieldColumnMap() as $_field => $_column) {
```
Would it make more sense to de-couple this method from field-column map awareness? Also, why is it in the same trait? Looks quite generic to me.

mecha replied:
To be honest, I've thought about extracting it from this trait.

This method is just another case of separating for the sake of cleanliness. I could have just had the body of this method go inside the `foreach` in line 76.

The reason why I didn't extract this was probably due to the fact that the other traits do not need this functionality - i.e. it's specific to inserting.

As for decoupling from field-column awareness ... I'm not sure what else it could depend on. This is THE method that changes fields to columns for insertion. Other traits have other means of doing this - some provide aliases to expression renderers, for example.

XedinUnknown replied:
> As for decoupling from field-column awareness ... I'm not sure what else it could depend on.

It can depend on field-column map, but one that is passed, not one that this instance is aware of. This would make it generic enough to go into a trait of its own.

mecha replied:
Oh! That makes sense.

XedinUnknown commented on src/InsertCapablePdoTrait.php:
```diff
@@ -0,0 +1,236 @@
+     */
+    protected function _insert($records)
+    {
+        $processedRecords = $this->_preProcessRecords($records, $valueHashMap);
```
Explain why this is necessary. AFAIU, it turns each key-value mapping in the record into key-valhash. But from here, this is not visible.

mecha replied:
Perhaps rename and redoc the method? I remember struggling to name this method, and so just called the procedure "pre-processing". If so, any suggestions?

XedinUnknown commented on src/InsertCapablePdoTrait.php:
```diff
@@ -0,0 +1,236 @@
+        $processedRecords = $this->_preProcessRecords($records, $valueHashMap);
+
+        $query = $this->_buildInsertSql(
+            $this->_getSqlInsertTable(),
```
What happens if an entity spans multiple tables?

mecha replied:
Don't know. Cannot insert into multiple tables in a single SQL statement. So that's probably responsibility that should exist at a higher level of abstraction.

What I would probably do is have a composite resource model, that understands what the entity is made of, split them accordingly and delegate separate queries to the different children resource models.

That's just what I would do. In any case, this is a trait for providing generic PDO insertion functionality. And insertions cannot handle multiple tables.

XedinUnknown replied:
It's true that insertion cannot be done to multiple tables in one query. But whether one or many - that can be abstracted. We treat CQRS resource models as something that is aware of an entity. An entity may span multiple tables. If it can be retrieved from multiple tables, then shouldn't it be also insertable to multiple tables, etc? Otherwise, how do you retrieve a multi-table entity, and then make a change to it?

mecha replied:
I agree with you. But this trait is not a trait for a resource model. It's just functionality for inserting records into the database. A higher level trait or class should handle something like this. Do you not agree?

XedinUnknown commented on src/InsertCapablePdoTrait.php:
```diff
@@ -0,0 +1,236 @@
+     *
+     * @return array The pre-processed record data list, as an array of record data associative sub-arrays.
+     */
+    protected function _preProcessRecords($records, &$valueHashMap = [])
```
Since this specifically converts key-value mappings into key-valhash mappings, does it make sense for this method to have such a generic name? This name hints to me that it maybe normalizes some values, not that it entirely changes them - inserting value hashes into a table wouldn't make much sense.

mecha replied:
Pretty much what I said [here](https://github.com/RebelCode/pdo-cqrs-resource-models/pull/1#discussion_r171797876). Any suggestions for the name?

Can you clarify what you meant by "inserting value hashes into a table wouldn't make much sense."? Is it because you disagree or don't understand why values are changed into hashes?

XedinUnknown replied:
I understand why they are changed into hashes. But the typical usage algo goes like this:

```
1. Receive records.
2. Do something with them (pre-process).
3. Insert each record.
```

You kind of expect the record to be inserted, i.e. retain its field-value relationship, albeit perhaps with some normalization. But here, if you just do that, then you will insert value hashes instead of values. That's why "pre-process" doesn't make sense: the result of pre-processing loses its meaning.

I would say that a good approach could be to agree on a term that would mean "replace some parts with placeholders, and record their mappings". Essentially, this is the opposite of interpolation, but also is interpolation too. A word like "surrogate".

mecha replied:
I understand. How about something a bit simpler than "surrogate"? Like "replace" or "substitute"? Currently leaning towards `_substituteRecordValuesForHashes()`.

XedinUnknown commented on src/PdoDeleteResourceModel.php:
```diff
@@ -0,0 +1,211 @@
+     * @param string|Stringable     $table              The name of the table from which records will be deleted.
+     * @param string[]|Stringable[] $fieldColumnMap     A map of field names to table column names.
+     */
+    public function __construct(PDO $pdo, TemplateInterface $expressionTemplate, $table, $fieldColumnMap)
```
Any missing throws, such as for invalid table or map?

mecha replied:
Yep, exactly the ones you mentioned. 👍

XedinUnknown reviewed (changes requested):
Overall, the problems that I've found are mostly missing documentation such as `@throws`, and explanation of why certain things are in a particular way.

mecha committed:
Synced PHPS dependency info with Composer

mecha committed:
Now wrapping string normalization exception

mecha committed:
Now wrapping internal exceptions with ...

... `OutOfRangeException`

mecha committed:
Fixed abstract uses of `_getPdoValueHashString`

mecha committed:
Fixed type docs for records and changesets

mecha committed:
Added missing constructor exceptions

mecha committed:
Removed unused `use` statement

XedinUnknown commented on src/InsertCapablePdoTrait.php:
```diff
@@ -26,7 +26,11 @@
      * @since [*next-version*]
      *
-     * @param array[]|ArrayAccess[]|stdClass[]|ContainerInterface[]|Traversable $records A list of records to insert.
+     * @param array|Traversable $records A list of records data containers. Accepted container types are:
```
This is not accurate. The list can also be an `stdClass`. Also, the previous doc used to reflect the types in a structured way, and therefore allowed static analysis. The new doc will not allow static analysis tools to understand whether the list contains valid items.

XedinUnknown committed:
Switched to newer deps or their newer versions

XedinUnknown committed:
Brought package up to date

- Fixed compatibility with newer version of abstract SQL package.
- Corrected tests.
- Now compatible with newer interfaces.

XedinUnknown committed:
Merge branch 'develop' of https://github.com/RebelCode/pdo-cqrs-resource-models into task/initial-classes

Conflicts:
	composer.json

XedinUnknown committed:
Now buiding in newer environment, and also on PHP 7.2
